x-pack/plugins/cases/server/client/cases/find.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import Boom from '@hapi/boom';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { fold } from 'fp-ts/lib/Either';
+import { identity } from 'fp-ts/lib/function';
+
+import {
+  CasesFindResponse,
+  CasesFindRequest,
+  CasesFindRequestRt,
+  throwErrors,
+  caseStatuses,
+  CasesFindResponseRt,
+  excess,
+} from '../../../common/api';
+
+import { createCaseError } from '../../common/error';
+import { constructQueryOptions, getAuthorizationFilter } from '../utils';
+import { includeFieldsRequiredForAuthentication } from '../../authorization/utils';
+import { Operations } from '../../authorization';
+import { transformCases } from '../../common';
+import { CasesClientArgs } from '..';
+
+/**
+ * Retrieves a case and optionally its comments and sub case comments.
+ *
+ * @ignore
+ */
+export const find = async (
+  params: CasesFindRequest,
+  clientArgs: CasesClientArgs
+): Promise<CasesFindResponse> => {
+  const { savedObjectsClient, caseService, authorization: auth, auditLogger, logger } = clientArgs;
+
+  try {
+    const queryParams = pipe(
+      excess(CasesFindRequestRt).decode(params),
+      fold(throwErrors(Boom.badRequest), identity)
+    );
+
+    const {
+      filter: authorizationFilter,
+      ensureSavedObjectsAreAuthorized,
+      logSuccessfulAuthorization,
+    } = await getAuthorizationFilter({
+      authorization: auth,
+      operation: Operations.findCases,
+      auditLogger,
+    });
+
+    const queryArgs = {
+      tags: queryParams.tags,
+      reporters: queryParams.reporters,
+      sortByField: queryParams.sortField,
+      status: queryParams.status,
+      caseType: queryParams.type,
+      owner: queryParams.owner,
+    };
+
+    const caseQueries = constructQueryOptions({ ...queryArgs, authorizationFilter });
+    const cases = await caseService.findCasesGroupedByID({
+      soClient: savedObjectsClient,
+      caseOptions: {
+        ...queryParams,
+        ...caseQueries.case,
+        searchFields:
+          queryParams.searchFields != null
+            ? Array.isArray(queryParams.searchFields)
+              ? queryParams.searchFields
+              : [queryParams.searchFields]
+            : queryParams.searchFields,
+        fields: includeFieldsRequiredForAuthentication(queryParams.fields),
+      },
+      subCaseOptions: caseQueries.subCase,
+    });
+
+    ensureSavedObjectsAreAuthorized([...cases.casesMap.values()]);
+
+    const [openCases, inProgressCases, closedCases] = await Promise.all([
```

Comment:
@jonathan-buttner Maybe we should check if we do it elsewhere and improve it as @legrego suggests?

Reply:
Yeah there's probably other places that we need to handle this. @legrego would it be ok if we created a new PR on top of this one to address the `Promise.all` issues or would you prefer that we did it in this PR?

Reply:
I'm not opposed to handling this in a followup. Let's just make sure this doesn't fall through the cracks, as this can cause issues for not just cases, but for all of Kibana

Reply:
Sounds good, I added it to this issue to keep track of the work: https://github.com/elastic/kibana/issues/100804